Libraries/Network/RCTNetworking.mm

```diff
@@ -333,22 +333,60 @@ - (RCTURLRequestCancellationBlock)processDataForHTTPQuery:(nullable NSDictionary
   return callback(nil, nil);
 }
 
-+ (NSString *)decodeTextData:(NSData *)data fromResponse:(NSURLResponse *)response
++ (NSString *)decodeTextData:(NSData *)data fromResponse:(NSURLResponse *)response withCarryStorage:(NSMutableData*)carryStorage
 {
   NSStringEncoding encoding = NSUTF8StringEncoding;
   if (response.textEncodingName) {
     CFStringEncoding cfEncoding = CFStringConvertIANACharSetNameToEncoding((CFStringRef)response.textEncodingName);
     encoding = CFStringConvertEncodingToNSStringEncoding(cfEncoding);
   }
+	
+	BOOL wasCarryStorageProvided = carryStorage != nil;
+	
+	if(carryStorage == nil)
+	{
+		carryStorage = [NSMutableData new];
+	}
+	
+  [carryStorage appendData:data];
+	
   // Attempt to decode text
-  NSString *encodedResponse = [[NSString alloc] initWithData:data encoding:encoding];
-  if (!encodedResponse && data.length) {
-    // We don't have an encoding, or the encoding is incorrect, so now we try to guess
-    [NSString stringEncodingForData:data
-                    encodingOptions:nil
-                    convertedString:&encodedResponse
-                usedLossyConversion:NULL];
-  }
+  NSString *encodedResponse = [[NSString alloc] initWithData:carryStorage encoding:encoding];
+	
+  NSMutableData* newCarryStorage = [NSMutableData new];
+	
+  if(!encodedResponse && data.length > 0)
+  {
+		if(wasCarryStorageProvided)
+		{
+			//If decode failed, we attempt to trim broken character bytes from the data.
+			CFStringEncoding cfEncoding = CFStringConvertNSStringEncodingToEncoding(encoding);
+			CFIndex maxCharLength = CFStringGetMaximumSizeForEncoding(1, cfEncoding);
+			
+			while(carryStorage.length > 0 && newCarryStorage.length < maxCharLength)
+			{
+				[newCarryStorage replaceBytesInRange:NSMakeRange(0, 0) withBytes:[carryStorage subdataWithRange:NSMakeRange(carryStorage.length - 1, 1)].bytes length:1];
+				carryStorage.length -= 1;
```

Comment:
instead of moving 1 byte for each iteration in the loop, why not use a local variable starting with `carryStorage.length`, then decrement that in the loop, then move the bytes around after the while loop. That's probably more efficient